Add catalog API documentation (#4023)
sensu/sensu-docs: content/sensu-go/6.8/catalog/catalog-reference.md

Sensu then applies the user's customizations to the integration's resource definitions and automatically deploys the integration configuration to agents in real time.
No external configuration management is required.

## Catalog repository example

The repository that stores Sensu integrations must organize files in the following structure:

{{< code text >}}
integrations/
└── <namespace>/
    └── <integration_name>/
        ├── img/
        │   ├── dashboard-1.gif
        │   └── dashboard-2.png
        ├── CHANGELOG.md
        ├── README.md
        ├── logo.png
        ├── sensu-integration.yaml
        └── sensu-resources.yaml
    └── <integration_name>/
        ├── img/
        │   ├── dashboard-1.gif
        │   └── dashboard-2.png
        ├── CHANGELOG.md
        ├── README.md
        ├── logo.png
        ├── sensu-integration.yaml
        └── sensu-resources.yaml
└── <namespace>/
    └── <integration_name>/
        ├── img/
        │   ├── dashboard-1.gif
        │   └── dashboard-2.png
        ├── CHANGELOG.md
        ├── README.md
        ├── logo.png
        ├── sensu-integration.yaml
        └── sensu-resources.yaml

> {{< /code >}}

Author: I think this section should include a note that explains that in the context of catalog integration organization, namespace does not mean the Sensu RBAC namesapce.

  server    Serves static catalog API for development purposes
  preview   Serves static catalog API & preview catalog web application for development purposes

FLAGS
  -integrations-dir-name integrations  path to the directory containing namespaced integrations
  -log-level info                      log level of this command ([panic fatal error warn info debug trace])
  -repo-dir .                          path to the catalog repository
{{< /code >}}

### Catalog versions

The catalog-api tool generates builds into a checksum-based output directory structure.
The `version.json` file manages the path to the latest or production catalog API content and instructs the web UI to load catalog contents from the specified checksum directory.
When you run the command to generate the catalog, catalog-api creates the `version.json` file.

The contents of a `version.json` file are similar to this example:

{{< code text "JSON" >}}
{
  "release_sha256": "5029648381dff2426ea247147456b4f1227fd6d9050fa42f0660e67a218f8c87",
  "last_updated": 1655840571
}
{{< /code >}}

If you make any changes to your integration files, the catalog-api tool will generate a new checksum directory.

> To revert to an older build, change the `release_sha256` in `version.json` to point to a different release directory.

Author: How do users change the release_sha256 in version.json?

Reviewer: If a user needs to revert to an older Catalog API release then they can open the version.json file, replace the existing value for the `release_sha256` field with the sha256 of the Catalog API they'd like to serve, and save it.

*Example:*
The current `version.json` file can be found by visiting `https://catalog.sensu.io/version.json`. This file tells the Web UI where to find the latest catalog release.

The current `release_sha256` checksum is `d0e2ba810c3d546c82121406fb5f214e66aeb7fe9026706f9a7391463cf4da19`. The Web UI can now construct the URL to retrieve that release of the catalog: https://catalog.sensu.io/d0e2ba810c3d546c82121406fb5f214e66aeb7fe9026706f9a7391463cf4da19/v1/catalog.json

If a previous release is contained within a directory with a sha256 checksum of `a0070d2c03de050c4ff116a33a675230a38193edac66862c7d2b1526caf91059` then the `release_sha` field in the `version.json` file can be replaced with that checksum.

### Catalog tags

The catalog-api tool consumes git tags and parses integration-specific git tags to manage and generate versioned integrations.

> This makes it possible to give users access to earlier versions of integrations to hedge against regressions in individual integrations.

Author: How do users publish multiple versions of an integration?

Reviewer: Users can tag an integration with a name matching `<integration_namespace>/<integration_filename>/<YYYYMMDD>.0.0` when an integration is ready to be released. The catalog-api tool will find all git tags matching that format to build the catalog. Multiple versions of one integration would require multiple git tags.

For example, in the [official Sensu Catalog repository][30], two `ansible-tower-remediation` version tags are defined:

{{< code text >}}
git tag --list |grep ansible-tower-remediation
ansible/ansible-tower-remediation/20220223.0.0
ansible/ansible-tower-remediation/20220421.0.0
{{< /code >}}

Using these tags, the catalog-api tool will generate the following integration structure:

{{< code text >}}
tree /tmp/generated-api/ -L 7
/tmp/generated-api/
├── release
│   ├── 5029648381dff2426ea247147456b4f1227fd6d9050fa42f0660e67a218f8c87
│   │   └── v1
│   │ ├── ansible
│   │ │   ├── ansible-tower-remediation
│   │ │   │   ├── 20220223.0.0
│   │ │   │   │   ├── CHANGELOG.md
│   │ │   │   │   ├── img
│   │ │   │   │   ├── logo.png
│   │ │   │   │   ├── README.md
│   │ │   │   │   └── sensu-resources.json
│   │ │   │   ├── 20220223.0.0.json
│   │ │   │   ├── 20220421.0.0
│   │ │   │   │   ├── CHANGELOG.md
│   │ │   │   │   ├── img
│   │ │   │   │   ├── logo.png
│   │ │   │   │   ├── README.md
│   │ │   │   │   └── sensu-resources.json
│   │ │   │   ├── 20220421.0.0.json
│   │ │   │   └── versions.json
│   │ │   └── ansible-tower-remediation.json
{{< /code >}}

> ## Use the Sensu Catalog API server during integration development

Author: I was unsuccessful at testing this process, but I'm not sure what is wrong with the instructions here.

When I try to test (using the official catalog repo), everything is OK until step 6. Then, I get:

```
[vagrant@sensu-centos catalog-api]$ ../catalog-api/catalog-api catalog server --repo-dir . -watch
2:23PM WRN Skipping integration version reason="unmatched git tag" tag=0.1.0-alpha.1
2:23PM WRN Skipping integration version reason="unmatched git tag" tag=0.1.0-alpha.10
2:23PM WRN Skipping integration version reason="unmatched git tag" tag=0.1.0-alpha.11
2:23PM WRN Skipping integration version reason="unmatched git tag" tag=0.1.0-alpha.12
2:23PM WRN Skipping integration version reason="unmatched git tag" tag=0.1.0-alpha.13
2:23PM WRN Skipping integration version reason="unmatched git tag" tag=0.1.0-alpha.14
2:23PM WRN Skipping integration version reason="unmatched git tag" tag=0.1.0-alpha.2
2:23PM WRN Skipping integration version reason="unmatched git tag" tag=0.1.0-alpha.3
2:23PM WRN Skipping integration version reason="unmatched git tag" tag=0.1.0-alpha.4
2:23PM WRN Skipping integration version reason="unmatched git tag" tag=0.1.0-alpha.5
2:23PM WRN Skipping integration version reason="unmatched git tag" tag=0.1.0-alpha.6
2:23PM WRN Skipping integration version reason="unmatched git tag" tag=0.1.0-alpha.7
2:23PM WRN Skipping integration version reason="unmatched git tag" tag=0.1.0-alpha.8
2:23PM WRN Skipping integration version reason="unmatched git tag" tag=0.1.0-alpha.9
error: error loading integrations: error retrieving integrations directory listing: open integrations: no such file or directory
[vagrant@sensu-centos catalog-api]$ ../catalog-api/catalog-api catalog server --repo-dir .
```

Reviewer: It looks like you're running the command from the wrong directory. You need to clone the catalog repository outside of the catalog-api directory and run the command from there.

api_version  | 
-------------|------ 
description  | Sensu API group and version for the resource to patch.
required     | true
type         | String
example      | {{< code yml >}}
api_version: core/v2
{{< /code >}}

name         | 
-------------|------ 
description  | Name of the resource to patch.
required     | true
type         | String
example      | {{< code yml >}}
name: nginx-metrics
{{< /code >}}

type         | 
-------------|------ 
description  | Type of the resource to patch.
required     | true
type         | String
example      | {{< code yml >}}
type: CheckConfig
{{< /code >}}

## Resource limits

There is no limit on the number of resources you can bundle into a single integration.
Each integration can include as many checks, event filters, handlers, and pipelines as you need to achieve your observability goals.
For example, you can develop a single host monitoring integration that includes all of the checks you want to run on every server.

## Check guidelines

For integrations that create checks, list the resource definitions in your `sensu-resources.yaml` file in in the following order:

1. CheckConfig
2. HookConfig
3. Secret
4. Asset

> The prompts for check resources should include at least one subscription.

Author: Caleb's catalog notes at https://github.com/sensu/catalog#sensu-integration-guidelines-1 say this is a requirement, but there are several integrations in the official catalog that do not include or require a subscription (the checks would require one to run obviously, but the integration workflow itself does not require providing a subscription to install the integration).

For this reason, I softened the language to "should" here, but I want to make sure that is accurate.

Reviewer: Whether that subscription is hardcoded into the integration or requested via a prompt, a check resource needs at least one subscription to be created. I think using "should" in the context of prompts is ok.

Subscriptions should be named according to the check's function.
For example, a PostgreSQL monitoring check could include a subscription named `postgresql`.

Use the YAML `>-` multiline block scalar syntax to wrap the check `command` value and make it easier to read.
For example:

{{< code yml >}}
spec:
  command: >-
    check-disk-usage.rb
    -w {{ .annotations.disk_usage_warning | default 85 }}
    -c {{ .annotations.disk_usage_critical | default 95 }}
{{< /code >}}

Use tunables like [tokens][25] in your check commands as needed, sourced from entity annotations (not labels) and with explicitly configured default values.

Check resources should use [interval scheduling][26] with a minimum interval of 30 seconds.
Set check `timeout` to a non-zero value that is no greater than 50% of the interval.

> Prompts for check pipelines should use one of the following generic categories:

Author: Is there a reason for always using one of these categories? Is this just for standardization among integrations that are part of the official Sensu Catalog?

Reviewer: Correct. These are the standard categories we chose to describe the types of checks that can exist. I believe we use this information in the Web UI for filtering/searching for integrations.

- Alerts
- Incident management
- Metrics
- Events
- Deregistration
- Remediation

## Pipeline guidelines

For integrations that create pipelines, list the resource definitions in your `sensu-resources.yaml` file in in the following order:

1. Pipeline
2. Handler, SumoLogicMetricsHandler, and TCPStreamHandler
3. Filter
4. Mutator
5. Secret
6. Asset

> For alert and incident management pipelines, do not use event filters that have subjective configuration options.

Author: Caleb's notes actually say "very subjective" -- can we be explicit about what we mean by subjective here? For example, do we mean something like "too restrictive" or "options that might not be clear to users" or something else?

Reviewer: I'm not exactly sure but my best guess is that they shouldn't be configured in a way that makes too many assumptions. cc @jspaleta

Author: Thank you. I asked in Slack https://sumologic.slack.com/archives/C0468FZ731C/p1666622236526099

        - path: /spec/output_metric_thresholds/1/thresholds/0/max
          op: replace
          value: "[[nginx_waiting_warn]]"
  post_install:
    - type: section
      title: Success
    - type: markdown
      body: |
        You enabled the NGINX Monitoring integration.
        The `nginx-metrics` check will run for all Sensu agents with these subscriptions: [[subscriptions]].
{{< /code >}}

Sensu Catalog integration definitions resemble other Sensu resources, but Sensu Go does not process them directly.
Instead, the [Sensu Catalog API][2] uses integration definitions along with the other files in the catalog repository, like READMEs and dashboard images, to generate a static API.

## Private catalogs

The [Sensu Catalog API][2] renders static HTTP API content that the Sensu web UI can consume.
This means you can create a private enterprise catalog of custom integrations and make it available to users in the Sensu web UI.

Consider forking the official Sensu Catalog repository, https://github.com/sensu/catalog, as a starting point for building your own private catalog.

The Catalog API defines integrations globally rather than by namespace.

> When you create a private catalog, all integrations in your repository are available for all users across namespaces in the web UI.

Author: Is this correct? Caleb said this in an earlier demo, but I believe Jef's more recent demo mentions something about controlling user access by namespace.

Reviewer: I thought this was true; cc @jspaleta.

Read [Build a private catalog of Sensu integrations][17] for more information.

## catalog-api command line interface tool

{{% notice note %}}
**NOTE**: The catalog-api tool is an alpha feature and may include breaking changes.
{{% /notice %}}

Sensu's [catalog-api][16] command line interface (CLI) tool uses the [Sensu Catalog API][2] to convert integration files into static API content that you can host on any HTTP web service.

Use the catalog-api tool to [generate a local catalog API][18] for testing as you develop new integrations and to build and run a private catalog.
Integration files must be stored in a repository that follows the required [organizational framework][16].

The catalog-api tool is written in Go.

### catalog-api subcommands

The catalog-api tool provides the following subcommands:

{{< code text >}}
catalog-api catalog --help
USAGE
  catalog-api catalog [flags] <subcommand> [flags]

SUBCOMMANDS
  generate  Generate a static catalog API
  validate  Validate a catalog directory and its integrations
  server    Serves static catalog API for development purposes
  preview   Serves static catalog API & preview catalog web application for development purposes

> 

Author: What is the difference between the server and preview subcommands?

Reviewer: The server subcommand simply starts a webserver to serve the catalog api (just a bunch of JSON files for the most part). In order to test the catalog while running the server subcommand, a user needs to set up a Sensu Backend and create a CatalogConfig resource to point to the webserver.

The preview subcommand does the same thing but also serves a preview dashboard that can communicate with Sensu Backend for easy development & testing of integrations (screenshot below). In this scenario, a user doesn't need to create a CatalogConfig resource or interact with the Sensu Backend's Web UI.

<img width="1212" alt="Screen Shot 2022-10-17 at 11 27 08" src="https://user-images.githubusercontent.com/120659/196255335-1dbe7e0d-7c8c-40b0-8318-a1404b421407.png">

To do this, use the `server` subcommand in the catalog-api command line tool.

{{% notice note %}}
**NOTE**: Make sure you have a local Sensu instance running with access to the Sensu web UI.
{{% /notice %}}

1. Clone the Sensu Catalog API repository:
{{< code shell >}}
git clone https://github.com/sensu/catalog-api
{{< /code >}}

2. Navigate to the local catalog-api repository:
{{< code shell >}}
cd catalog-api
{{< /code >}}

3. Build the `catalog-api` tool:
{{< code shell >}}
go build
{{< /code >}}

4. Clone the repository that stores your Sensu integrations.
This example uses Sensu's public integration repository:
{{< code shell >}}
git clone https://github.com/sensu/catalog
{{< /code >}}

5. Navigate to your local copy of the repository that stores the Sensu integrations.
This example uses https://github.com/sensu/catalog, so the repository is `catalog`: 
{{< code shell >}}
cd ../catalog
{{< /code >}}

6. Run the catalog-api `server` subcommand:
{{< code shell >}}
../catalog-api/catalog-api catalog server --repo-dir . -watch
{{< /code >}}

   The `.` in the command tells Sensu to read the catalog contents from your local environment.

>    Use the `-watch` flag to reload the API as you save updates in integration files so that you can see them live in the Sensu web UI.

Author: I'm not sure if this command is correct

Author: There seems to be another step here about adding two headers: one to override CORS and the other to prevent caching the browser content. Jef does this with Python in the customer demo, but I'm not sure how users should do this.

Reviewer: The server in the catalog-api tool deals with CORS & caching for us. Users will only need to set them if they're serving the catalog api using a different webserver.

Author: Can you expand on this a bit? I'm not sure what you mean by a different webserver. Jef showed it in a customer demo, so I'm thinking we have at least one customer who wants or needs to do this.

Reviewer: The catalog-api tool can perform two main tasks:
1. It can look at a Catalog Git repository & generate a Catalog API. A Catalog API is a hierarchical structure of directories, JSON files, YML files, and images. These directories & files are useless by themselves; they need a web server to give the world (or other users) access to them through HTTP/HTTPS.
2. A web server that serves the generated Catalog API.

The web server built into the catalog-api tool should only be used for local development. Users should use a solution such as Nginx, Apache, Amazon S3/Cloudfront, etc. The instructions on configuring CORS and caching is dependent on which software the user is using to serve the Catalog API.

Author: Ah ok, that makes sense. Thank you.

---
title: "Catalog integrations reference"
linkTitle: "Catalog Integrations Reference"
reference_title: "Catalog integrations"
type: "reference"
description: "Use Sensu's Catalog API to build a private catalog of Sensu integrations. Read the reference to create integrations for your catalog."
weight: 80
version: "6.8"
product: "Sensu Go"
menu: 
  sensu-go-6.8:
    parent: catalog
---

{{% notice commercial %}}
**COMMERCIAL FEATURE**: Access the Sensu Catalog and integrations in the packaged Sensu Go distribution.
For more information, read [Get started with commercial features](../../commercial/).
{{% /notice %}}

{{% notice note %}}
**NOTE**: The Sensu Catalog is in public preview and is subject to change.

> {{% /notice %}}

Author: Is this still true? We have this note on our existing catalog page (https://docs.sensu.io/sensu-go/latest/web-ui/sensu-catalog/).

Reviewer: I'm not sure about this. cc @portertech @fguimond

The [Sensu Catalog][1] is a collection of Sensu integrations that provide reference implementations for effective observability.

> The contents of the official Sensu Catalog are periodically published to the official Sensu Catalog API, which is hosted at https://catalog.sensu.io and displayed within the Sensu web UI.

Author: ```suggestion
The contents of the official Sensu Catalog are periodically published with the Sensu Catalog API, which is hosted at https://catalog.sensu.io and displayed within the Sensu web UI.
```

When users install integrations in the Sensu web UI, they receive prompts to enter information.
For example, the DNS Monitoring integration includes prompts for the domain name, record type, record class, servers, and port to query.
Sensu then applies the user's customizations to the integration's resource definitions and automatically deploys the integration configuration to agents in real time.
No external configuration management is required.

> 

Author: ```suggestion
The Sensu Catalog provides a way for you and your teams to configure powerful real-time monitoring and observability for the systems you rely on.
Integrations are self-service, and the Catalog is designed to help you scale up with fewer barriers.

```

  server    Serves static catalog API for development purposes
  preview   Serves static catalog API & preview catalog web application for development purposes

FLAGS
  -integrations-dir-name integrations  path to the directory containing namespaced integrations
  -log-level info                      log level of this command ([panic fatal error warn info debug trace])
  -repo-dir .                          path to the catalog repository
{{< /code >}}

The server subcommand starts a webserver to serve the JSON files the Catalog API generates.
To view your catalog in the web UI while running the server subcommand, you must also configure a Sensu backend and create a GlobalConfig resource to point to the webserver.

The preview subcommand starts a webserver like the server subcommand but also serves a preview web UI that can communicate with the Sensu backend.
If you use the preview subcommand, you do not need to interact with the Sensu backend or create a GlobalConfig resource.
The last line of the preview subcommand response provides the address to use to view the preview catalog in your browser.
For example:

{{< code text >}}
10:07AM INF API server started address=:3003

> {{< /code >}}

Author: I'm running into a couple confusing things about the server and preview commands.

Both server and preview result in the `INF API server started address=:3003` response. But only preview actually generates anything at localhost:3003 (if you run server and then try to visit localhost:3003, it's a 404). Is this intentional? Maybe I'm misinterpreting the server started response.

If I delete integrations from my local copy of the catalog repo and then run the preview subcommand, the integrations still appear in the web UI. I can't figure out how this is happening, unless it's because I ran validate before deleting the integrations.

Reviewer: Strange, you shouldn't be seeing a 404 in server mode. Let's have a call to figure out why that's not working for you. Caching could be affecting the second issue but I can take a look on the call.

Author: Resolved in a call w/Justin

For example, a PostgreSQL monitoring check could include a subscription named `postgresql`.

Use the YAML `>-` multiline block scalar syntax to wrap the check `command` value and make it easier to read.
For example:

{{< code yml >}}
spec:
  command: >-
    check-disk-usage.rb
    -w {{ .annotations.disk_usage_warning | default 85 }}
    -c {{ .annotations.disk_usage_critical | default 95 }}
{{< /code >}}

Use tunables like [tokens][25] in your check commands as needed, sourced from entity annotations (not labels) and with explicitly configured default values.

Check resources should use [interval scheduling][26] with a minimum interval of 30 seconds.
Set check `timeout` to a non-zero value that is no greater than 50% of the interval.

Prompts for check pipelines should use one of the following generic categories:

- Alerts
- Incident management
- Metrics
- Events
- Deregistration
- Remediation

## Pipeline guidelines

For integrations that create pipelines, list the resource definitions in your `sensu-resources.yaml` file in the following order:

1. Pipeline
2. Handler, SumoLogicMetricsHandler, and TCPStreamHandler
3. Filter
4. Mutator
5. Secret
6. Asset

> For alert and incident management pipelines, do not use event filters that have subjective configuration options.
> Use the built-in [is_incident][27] and [not_silenced][28] filters.

Author: ```suggestion
For alert and incident management pipelines, we recommend using the built-in [is_incident][27] and [not_silenced][28] event filters instead of custom event filters that are configured for specific use cases.
```

### catalog-api subcommands

The catalog-api tool provides the following subcommands.
Replace <REPO_NAME> with the name of your local repository of integrations:

{{< code text >}}
catalog-api <REPO_NAME> --help

USAGE

>   catalog-api <REPO_NAME> [flags] <subcommand> [flags]

Reviewer: I think the usage was previously correct here. There aren't any arguments for a repo name but there is a flag for repo dir (`--repo-dir path/to/catalog/repo`).

Author: Ahhh I see, I was getting turned around by using the official "catalog" repo and thought the command was pulling that. I'll fix this, thank you very much! 😅
